src/udp.rs

```diff
@@ -358,33 +362,29 @@ impl IsEmpty for (Vec<u8>, SocketAddr) {
     }
 }
 
+/// UDP socket receive result. It yields data only when socket returns EWOULDBLOCK.
+enum RecvResult<T> {
+    /// Stop reading socket and return some data, if any was read.
+    WouldBlock(Option<T>),
+    /// Keep reading UDP socket until it returns EWOULDBLOCK.
+    _ContinueRecv,
+}
+
+/// Buffers received UDP packets and checks when to stop calling `recv()`.
 fn handle_recv_res<T: IsEmpty>(
     recv_res: io::Result<T>,
     read_buffer: &mut VecDeque<T>,
-    would_block: &mut bool,
-) -> ::Res<Option<T>> {
+) -> ::Res<RecvResult<T>> {
     match recv_res {
         Ok(msg) => {
-            if msg.is_empty() {
-                let e = Err(SocketError::ZeroByteRead);
-                if !read_buffer.is_empty() {
-                    return match read_buffer.pop_front() {
-                        Some(msg) => Ok(Some(msg)),
-                        None => e,
-                    };
-                } else {
-                    return e;
-                }
+            if !msg.is_empty() {
+                read_buffer.push_back(msg);
             }
-            read_buffer.push_back(msg);
-            Ok(None)
+            Ok(RecvResult::_ContinueRecv)
```

Comment:
If you are using it then why have the leading `_` - that was more to prevent "not used" warnings.